### PR TITLE
🐛 Source Shopify: fix the connector's performance for `incremental refresh`

### LINKED ISSUE #5469: Source Shopify: native airbyte is slower than singer tap
## Tell us about the problem you're trying to solve
From user [slack convo](https://airbytehq.slack.com/archives/C01MFR03D5W/p1628600395445900):
> Hi all, the Shopify native connector takes a very long time to sync even in incremental mode due to it getting stuck on a retry loop. I’m attaching logs as a reply to this message
> hmm I have been running the singer-based connector for a few months without facing similar issues. Is it because it implements some of the enhancements you mentioned? Unfortunately I might not be able to adopt the native connector (I really want to) if it’s not at least on par with the Singer-based one.

## Describe the alternative you’ve considered or used
A clear and concise description of any alternative solutions or features you've considered or are using today.

## Are you willing to submit a PR?

Remove this with your answer :-)

### PR DESCRIPTION
**Pull request by bazarnov**
## What
#5469 - Source Shopify: native airbyte is slower than singer tap

## How
- Added the `utils.backup_stream_state` decorator function - to handle the slicing using the incremental refresh for parent stream.
- Added unit_test to cover the changes.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
 
- [X] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated 
    - [X] Connector's `README.md`
    - [X] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [X] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

### CODE REVIEW

**Review by antixar — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 1]
```diff
@@ -302,9 +342,9 @@ def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
...
         self, stream_state: Mapping[str, Any] = None, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs
     ) -> Iterable[Mapping[str, Any]]:
-        stream_state = stream_state or {}
-        price_rules_stream = PriceRules(authenticator=self.authenticator, shop=self.shop, start_date=self.start_date)
-        for data in price_rules_stream.read_records(sync_mode=SyncMode.full_refresh):
+        # get the last saved orders stream state
+        price_rules_stream_state = self.tmp_stream_state.get("price_rules")
+        for data in PriceRules(self.config).read_records(sync_mode=SyncMode.incremental, stream_state=price_rules_stream_state):
             discount_slice = super().read_records(stream_slice={"price_rule_id": data["id"]}, **kwargs)
             yield from self.filter_records_newer_than_state(stream_state=stream_state, records_slice=discount_slice)
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py]: Done.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 2]
```diff
@@ -124,23 +163,24 @@ def filter_records_newer_than_state(self, stream_state: Mapping[str, Any] = None
...
     def read_records(
         self, stream_state: Mapping[str, Any] = None, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs
     ) -> Iterable[Mapping[str, Any]]:
-        orders_stream = Orders(authenticator=self.authenticator, shop=self.shop, start_date=self.start_date)
-        for data in orders_stream.read_records(sync_mode=SyncMode.full_refresh):
+        # get the last saved orders stream state
+        orders_stream_state = self.tmp_stream_state.get("orders")
+        for data in Orders(self.config).read_records(stream_state=orders_stream_state, **kwargs):
             slice = super().read_records(stream_slice={"order_id": data["id"]}, **kwargs)
             yield from self.filter_records_newer_than_state(stream_state=stream_state, records_slice=slice)
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:180]: Done.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 3]
```diff
@@ -49,10 +49,18 @@ class ShopifyStream(HttpStream, ABC):
...
+    It's empty, once the sync has started and is being updated while sync operation takes place,
+    It holds the `temporary stream state values` before they are updated to have the opportunity to reuse this state.
+    """
+    tmp_stream_state = {}
+
+    def __init__(self, config: Dict):
+        super().__init__(authenticator=config["authenticator"])
+        self.start_date = config["start_date"]
+        self.shop = config["shop"]
+        self.config = config
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:62]: I realised that we don't need to declare those variables at all now, we have them inside of `self.config`. Done.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 4]
```diff
@@ -89,17 +97,48 @@ def data_field(self) -> str:
...
+    def stream_state_to_tmp(
+        self,
+        stream_name: str,
+        current_stream_state: MutableMapping[str, Any],
+        latest_record: Mapping[str, Any],
+        state_object: Mapping[str, Any],
+        cursor_field: str,
+    ) -> Mapping[str, Any]:
+        """
+        Method to save the current stream state for future reuse within slicing.
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py]: Done.

**Review by antixar — APPROVED** [review]
Great!

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py [hunk 5]
```diff
@@ -108,3 +109,53 @@ def wrapper_balance_rate_limit(*args, **kwargs):
             return wrapper_balance_rate_limit
 
         return decorator
+
+
+class BackupStreamState:
```
Comment (davinchia) [inline airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:114]: nit: can we change this to `EagerlyCachedStreamState`?

Backup sounds like we are backing something up. Here, what we are doing is making sure we eagerly using already loaded state.

**Review by davinchia — APPROVED** [review]

File: airbyte-integrations/connectors/source-shopify/source_shopify/utils.py (hunk 5 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/utils.py:114]: Corrected. Thank you)

**Pull request merged**